### PR TITLE
Texture atlas for mapblocks meshes

### PR DESCRIPTION
**Pull request by Andrey2470T**
This PR has a trial to decrease the drawcalls count in the clientmap, mainly switches count of the textures in the OpenGL state. The crucial changes which this PR implements:

- Collecting all textures from the tile layers during the media loading and drawing them into the one atlas texture.
- Using the single huge texture (atlas) for all mapblocks meshes and their submeshes.
- Improved batching the submeshes by the same material within the mapblock.
- Avoiding unnecessary setMaterial() calls.
- Allowing the atlases to be animateable (re-draw the frames of the animated tiles).

An example of the animated atlas generated by this PR in MT:
animated_atlas

Note, the atlasing is not applied in the mesh nodes as an exception for this drawtype, they use own textures spaces. There are many visual artifacts with some drawtypes, mainly UV issues, also filtering and mipmapping ones.

## Known issues (mostly small ones)

- Incorrect UVs of the `flowingliquid` drawtype nodes
- Sometimes mesh nodes whose UVs are beyond 0-1 range render on themselves an odd geometry parts
- With the filtering enabled the crack textures are rendered skewed
- With the filtering + mipmapping enabled the lower mip levels still get a pixel pollution
- The atlas is black on Android (reported by @AbduSharif)

## How to test
1. See how nodes placed in the world look with various drawtypes.
2. Enable/disable filtering, mipmapping and other graphics settings.

### CODE REVIEW

**Comment by sfan5** [comment]
Obligatory mention that we had texture atlasing years ago and we had reasons for removing it at the time.
someone should find out: What were the reasons at the time? How is this different? Or did the circumstances change?

Have you performance-tested the impact of your PR?

> Avoiding unnecessary setMaterial() calls.

FWIW this is mostly irrelevant. Most of the state inside `SMaterial` is only updated to GL lazily when it's needed.

> Allowing the atlases to be animateable (re-draw the frames of the animated tiles).

How is this implemented?
Do you duplicate the alias for each frame? That would explode in complexity quite quickly.
Or repeatedly modify the texture? I imagine that runs into problems for (de)synchronizing animations quite quickly.

File: irr/src/COpenGLCoreTexture.h [hunk 1]
```diff
@@ -422,6 +422,43 @@ class COpenGLCoreTexture : public ITexture
 		Driver->getCacheHandler()->getTextureCache().set(0, prevTexture);
 	}
 
+	void drawToSubImage(int x, int y, int width, int height, ITexture *texture) override
```
Comment (sfan5) [inline irr/src/COpenGLCoreTexture.h:425]: wrong abstraction for this function

calling `lock()` on an `ITexture` will download the pixel data back from the GPU, which is a pointless waste of time.
this should either
* do the copy on the gpu
* take an `IImage` instead

**Comment by rubenwardy** [comment]
> Obligatory mention that we had texture atlasing years ago and we had reasons for removing it at the time.
someone should find out: What were the reasons at the time? How is this different? Or did the circumstances change?

It was removed due to bugs - I remember a lot of issues with nodes being solid colors and glitchy rather than textured. See https://github.com/minetest/minetest/pull/789

However, texture atlasing is the right:tm: way to do things for performance so definitely worth having

**Comment by rubenwardy** [comment]
Make sure to test this with dreambuilder or something that nearly maxes out the node registration limit, with a lot of animated textures

**Comment by sfan5** [comment]
also a general thought: why not introduce a texture atlas for static nodes first and work on the rest later?
It seems you also changed the way vertices are batched, that would be better for a separate PR. (and what about transparency sorting?)

**Comment by Andrey2470T** [comment]
> Have you performance-tested the impact of your PR?
> 
I tested mainly at the `viewing_range=190` and `max_fps=60 with the NVIDIA GeForce GTX 1050 videocard, in MTG without/with mods as `homedecor`, `moreores`, `pipeworks` and a few other, but I can not confirm it increases the FPS and decreases the drawtime much. The switches of textures is one of the rendering bottleneck problem, so here the expensive frustum and occlusion cullings plays a big role in it, also a big count of the VBOs switches. This PR fixes only the texture switches related proble.
> FWIW this is mostly irrelevant. Most of the state inside SMaterial is only updated to GL lazily when it's needed.
> 
I wasn't sure on this account, I think there are GL functions (maybe the mix mode, face culling) which the SMaterial repeatedly calls if they are not changed from the buffer to buffer. The more exact control over it should ensure it won't happen in any case.

> How is this implemented? Do you duplicate the alias for each frame? That would explode in complexity quite quickly. Or repeatedly modify the texture? I imagine that runs into problems for (de)synchronizing animations quite quickly.
>
No, I don;'t certainly duplicate that. The atlas is always the same during the tile animations updates. Only redrawings happen onto the certain tiles which represent the animations frames if they should be replaced to the new frames. For this `ITexture::drawToSubImage()` is used.

**Comment by Andrey2470T** [comment]
> also a general thought: why not introduce a texture atlas for static nodes first and work on the rest later?
>
Because the atlas simplifies and also optimizes the tile animations updates. There's no a need to switch the frame textures of the same kind of animation for each animated tile which has it, only the redrawing to the single tile happens. And how my gif image shows in the first post it works successfully and fast.
> It seems you also changed the way vertices are batched, that would be better for a separate PR. (and what about transparency sorting?)
>
I changed it because all tile meshes has the same texture, therefore it has got possible to batch them within the mapblock still better than before.

**Comment by appgurueu** [comment]
I am not convinced that an atlas is the way to go. The problem of packing the textures - which can have arbitrary sizes - into an atlas alone already is nontrivial. There may be many examples where this works somewhat decently, but what about high resolution textures or extreme aspect ratios?

Now add to that things like animations. It appears you're basically duplicating the atlas for implementing animations? This is very wasteful in terms of texture memory: I can have a single animated texture and suddenly the texture space usage doubles.

Have you considered more modern alternatives such as [array textures](https://www.khronos.org/opengl/wiki/Array_Texture)? I believe these could make for a much more manageable, and possibly more performant, implementation on reasonably modern hardware (OGL 3.0+). Related: [Auri's video on the matter](https://github.com/AurailusVideos/SparseBindlessTextureArrays).

**Comment by Andrey2470T** [comment]
> I am not convinced that an atlas is the way to go. The problem of packing the textures - which can have arbitrary sizes - into an atlas alone already is nontrivial. There may be many examples where this works somewhat decently, but what about high resolution textures or extreme aspect ratios?
>
 The solution to that could be creating one more atlas if the current atlas size exceeds GL_MAX_TEXTURE_SIZE is what very unlikely. The minimum size that the texture can have is 16k in both dimensions, so it can fit `16 000 ^ 2 / 16 ^ 2 = 1'000'000` 16x textures. This is a huge amount and much more than enough and the most mods including the most popular games use them (MTG, VoxeLibre, NodeCore, Mineclonia, Exile and etc). If 128x texture pack is used the general texture count that could fit is `16^2 / 128 ^ 2 = 15625`, so it is still a high count. Higher-resolution texture packs just doesn't exist in MT yet.

> Now add to that things like animations. It appears you're basically duplicating the atlas for implementing animations? This is very wasteful in terms of texture memory: I can have a single animated texture and suddenly the texture space usage doubles.
> 
Where did you see I duplicate the atlas for animations?

> Have you considered more modern alternatives such as [array textures](https://www.khronos.org/opengl/wiki/Array_Texture)? I believe these could make for a much more manageable, and possibly more performant, implementation on reasonably modern hardware (OGL 3.0+).
> 
> Related: [Auri's video on the matter](https://github.com/AurailusVideos/SparseBindlessTextureArrays). She had to solve the same problem for her voxel game engine.
>
AFAIK the texture arrays are actually more limited than atlases. They can have fixed-size textures inside layers, the atlases can pack textures with arbitrary sizes. Also, Irrlicht doesn't support that kind of texture, so it is still an additional work.

**Comment by sfan5** [comment]
> Only redrawings happen onto the certain tiles which represent the animations frames if they should be replaced to the new frames.

So you upload (parts of) a texture between drawcalls. That sounds suboptimal for performance, are we just swapping one bottleneck for another?

> Because the atlas simplifies and also optimizes the tile animations updates. There's no a need to switch the frame textures [...]

I know there are technical reasons. What I mean is organisational reasons to split the PR into parts that are more easily tested, reviewed or bisected as one.

**Comment by Andrey2470T** [comment]
> So you upload (parts of) a texture between drawcalls. That sounds suboptimal for performance, are we just swapping one bottleneck for another?
> 
Re-drawing of the single animated tile in atlas is much faster than switching frame textures of each animated submesh in each visible mapblock which can be very many around the map. Also, theoretically it leads to the desyncronization if there are many such switches.

File: src/client/texture_atlas.cpp [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+
+		atlas_size += (tile_info.width * tile_info.height);
+	}
+
+	u32 atlas_side = std::min(getClosestPowerOfTwo(std::sqrt((f32)atlas_size)), 16384u);
+
+	packTextures(atlas_side);
+
+	// Extending the atlas width twice for allocating the additional texture space for cracks
+	m_atlas_texture = m_driver->addTexture(core::dimension2du(atlas_side*2, atlas_side), "DiffuseAtlas", video::ECF_A32B32G32R32F);
```
Comment (red-001) [inline src/client/texture_atlas.cpp:86]: this dimensions is a worst case bound right?  Could check the actual size of the packed texture and use that.

File: src/client/texture_atlas.cpp [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+{
+	return std::pow(2u, (u32)std::ceil(std::log2(num)));
+}
+
+bool TextureAtlas::canFit(const TileInfo &area, const TileInfo &tex)
+{
+	return (tex.width <= area.width && tex.height <= area.height);
+}
+
+void TextureAtlas::packTextures(int side)
```
Comment (red-001) [inline src/client/texture_atlas.cpp:109]: Could add a note here on what texture packing algorithm is used.

File: irr/src/COpenGLCoreTexture.h (hunk 1 above)
Comment (Andrey2470T) [reply irr/src/COpenGLCoreTexture.h:430]: I've fixed it just getting the first cached image from the new `ITexture::getImagesCache()` and drawing its data

File: src/client/texture_atlas.cpp [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+						return (area1.width*area1.height < area2.width*area2.height);
+					});
+				break;
+			}
+		}
+
+		if (!packed)
+			return;
+
+		counter++;
```
[the remark this answers is not in the document]
Comment (Andrey2470T) [reply src/client/texture_atlas.cpp]: Fixed.

File: src/client/texture_atlas.cpp (hunk 2 above)
Comment (Andrey2470T) [reply src/client/texture_atlas.cpp:87]: Fixed also.

Comment (red-001) [reply src/client/texture_atlas.cpp:87]: I was asking if `core::dimension2du(atlas_side*2, atlas_side)` is a worst case for the atlas dimensions, and suggesting that the actual max width and height is used (can be calculated in `packTextures`). 

I see the change made the `GL_MAX_TEXTURE_SIZE` worst case bound was replaced with the actual value, was I just mistaken about the atlas dimensions side of things? Will it always need exactly that size texture?


On a general note, I think it would be cleaner in any case if `packTextures` returned the bounds it needs. Just cause the bounds depend on *how* the textures are packed.

**Comment by red-001** [comment]
how does this interact with dynamic media loading at runtime?

**Comment by Desour** [comment]
One of the issues that texture atlases cause is that textures bleed into neighbors on the atlas if texture filters are used.
Take this as example:
screenshot_20240828_124809

There are workarounds where you put stuff between the textures (idk how well one can make it work in the end). But array textures are the proper thing to use nowadays, AFAIK.

---

(Just FYI, when testing, the get setting calls in the mesh collector were eating much of my cpu. Cache them.)

**Comment by Andrey2470T** [comment]
> One of the issues that texture atlases cause is that textures bleed into neighbors on the atlas if texture filters are used. Take this as example: screenshot_20240828_124809
> 
> There are workarounds where you put stuff between the textures (idk how well one can make it work in the end).
>
Yes, I know about that issue with the filtering. That happens mainly with the bilinear and trilinear ones. With the anisotropic filter the weird stripes are rendered in the distance. The solution to that is creating one-pixel size frames around each texture in the atlas with the duplicating the edge pixels. So it will be necessary a bit to extend the atlas size.

> But array textures are the proper thing to use nowadays, AFAIK.
>
I think the atlas for MT is enough for now. It is very unlikely the atlas texture will run out of the max size (see my calculations above) because the texture sizes are mostly 16 or 32pix used in mods. If one more atlas is necessary it will be easy to add it, e.g. NodeDefManager could save as a vector of those. And then TileLayers could reference not only to the own TextureInfo by an index, but also to the atlas in that vector.

> (Just FYI, when testing, the get setting calls in the mesh collector were eating much of my cpu. Cache them.)
>
Hmm, I didn't check the CPU load in the htop when testing my branch, but really is getting the `enable_shaders` setting pretty expensive?

**Comment by Andrey2470T** [comment]
> how does this interact with dynamic media loading at runtime?
>
Actually in no way. It draws into the atlas only those textures which loaded at the startup. But I think it is possible to rebuild the atlas if a new texture gets loaded in the texturesource at run-time.

**Comment by AbduSharif** [comment]
Tried this on Android, everything block is black.

**Comment by Andrey2470T** [comment]
> Tried this on Android, every block is black.

Can you elaborate on which graphic settings did you test on Android? Was mipmaps/filtering enabled?

**Comment by Andrey2470T** [comment]
Summing up which issues are left to fix in this PR:

- Enabled filtering (bilinear/trilinear/anisotropic).
- Meshes whose UVs get beyond 0-1 range sometimes render in themselves unrelated parts of the geometry with the noise like in this screenshot:
 Снимок экрана от 2024-09-11 23-39-00
- Fix UVs of the top faces of the `flowingliquid` drawtype meshes.

So, by now the most UV issues with various drawtypes were fortunately fixed, also the mipmapping artifacts were eliminated:
Снимок экрана от 2024-09-11 23-37-03

**Comment by AbduSharif** [comment]
> Can you elaborate on which graphic settings did you test on Android? Was mipmaps/filtering enabled?

Defaults, no and no.

File: src/client/texture_atlas.cpp (hunk 3 above)
Comment (Andrey2470T) [reply src/client/texture_atlas.cpp:149]: Added the comment in the last commit.

**Comment by AbduSharif** [comment]
> Tried this on Android, every block is black.

There are no errors btw.

**Comment by Andrey2470T** [comment]
> > Tried this on Android, every block is black.
> 
> There are no errors in the log btw.

If the nodes are black, it means the atlas is a black texture also. Probably some problem with the `glTexSubImage2D()` on the ogles.

**Comment by AbduSharif** [comment]
Btw, why are you using one atlas texture instead of separating it into multiple based on the object type (like in Minecraft, IIRC its few atlas textures)?

**Comment by AbduSharif** [comment]
To be more to the point, Android devices don't support the same max texture size, a device someone I know (which still runs Minetest fine) has 4k as the max, and the one I tested this on (my main) has 8k as the max, a more recent one has 16k...etc.

**Comment by Andrey2470T** [comment]
> Btw, why are you using one atlas texture instead of separating it into multiple based on the object type? (like in Minecraft, IIRC its few atlas textures based on the object type. IIRC Minecraft also redraws them when settings change, like the mipmapping ones...etc, [here's](https://minecraft.wiki/w/Texture_atlas) a link for the interested)
> 
I think adding atlases for mapblock meshes has a higher priority since the loaded clientmap can have hundreds of textures at the middle or even thousands depending on the count of the loaded mapblocks and diversity of the nodes. For the remain things like entities, particles, gui it can be done later. At the middle the entities and particles are not so many around the map as the nodes.

> Also what would be the max size of the atlas texture? It would be different based on the textures/texturepack right?
>
The maximal allowed OpenGL texture size is defined by the `GL_MAX_TEXTURE_SIZE` constant and depends on the platform (on my one it is 32k in both dimensions).

**Comment by Andrey2470T** [comment]
Also, my implementation takes into account the cases when the atlas size exceeds the max allowed one: it just adds a new atlas and does the independent packing of the remain textures in the passed array.

**Comment by AbduSharif** [comment]
I tested on another Android device, same black blocks issues.

**Comment by Desour** [comment]
Tested it a bit and looked a bit at code.
(I do not want to fully review right now or in the near future, sorry, but it's too messy, and the its fate is unclear.)

A map generated with https://codeberg.org/Desour/test_mapblock_drawcalls is fast now, so PR does what it wants to do.

Tried joining a server with a busy spawn (yourland test server), and the textures were all black.

If you want to avoid the texture bleeding issues when used with mipmapping, you could use the same texture size for all textures in the atlas, possibly by upscaling textures. (Would also have to be done for array textures.)

If I'm reading the code correct, you're dedicating half of the atlas to crack variants. That's a bit much.

I'd be interested in more high level descriptions of the implementation, btw.. Having to read through code to know what something does is quite cumbersome.

**Comment by Andrey2470T** [comment]
> Tested it a bit and looked a bit at code. (I do not want to fully review right now or in the near future, sorry, but it's too messy, and the its fate is unclear.)
>
I thought I had added enough thorough and descriptive comments throughout the atlas file, particularly about what is the TileInfo, which the packing algorithm used, how the filtering solved in the atlas and etc.

What do you mean "its fate is unclear"?
> A map generated with https://codeberg.org/Desour/test_mapblock_drawcalls is fast now, so PR does what it wants to do.
> 
It's nice to hear there's a confirmation this PR wins in the performance. Thanks for testing.
> Tried joining a server with a busy spawn (yourland test server), and the textures were all black.
>
Hmm, I had noticed the mips starting with the 1st level looked fully black when I joined any server and it didn't happen in the singleplayer mode. However, many changes were made since that moment and I need to check out it again.

> If I'm reading the code correct, you're dedicating half of the atlas to crack variants. That's a bit much.
>
Yes, that's a pretty terrible hack. I think about another implementation of that rendering the single node mesh containing only crack textures with the overlay mixing mode around the cracked node.

**Comment by Andrey2470T** [comment]
> I'd be interested in more high level descriptions of the implementation, btw.. Having to read through code to know what something does is quite cumbersome.
>
Would it simplify the reading the code for you if I added the short description about the implementation in the PR first post?

**Comment by Desour** [comment]
>What do you mean "its fate is unclear"?

AFAIK, there has been no clear coredev concept approval. I'm myself also not sure if we want to have a texture atlas. (The alternative would be to wait for someone who makes a PR for array textures.)
=> A review at this point would be a waste of my and your time, and it would wrongly suggest my concept approval.

>Would it simplify the reading the code for you if I added the short description about the implementation in the PR first post?

Yes!
I.e. what things were changed in map mesh generation, how you're managing the atlases, and so on.

**Comment by Andrey2470T** [comment]
> AFAIK, there has been no clear coredev concept approval. I'm myself also not sure if we want to have a texture atlas.
> 
There was an approval from rubenwardy in the beginning of the PR: https://github.com/minetest/minetest/pull/15061#issuecomment-2310276596
And which reasons could be of not having the optimisation in MT (taking into account the most of issues caused by the atlas were fixed in my last commits and now it's just required to do still a few ones also?).

**Comment by celeron55** [comment]
I think a working texture atlas implementation will benefit MT. I think the old one was removed because it interacted badly with various things.

As long as this is:
- benchmarked to confirm the benefit in performance, and
- tested to not cause artifacts that users will immediately start complaining after release,

I think this will be a good addition.

**Comment by lhofhansl** [comment]
How does this interact with `client_mesh_chunk`, if at all?

**Comment by Andrey2470T** [comment]
> How does this interact with `client_mesh_chunk`, if at all?

The same way as it interacted with that before. The PR just actually only change the general mapblock mesh structure and that how the prebuffers are merged within the mapblock.

**Comment by Desour** [comment]
@Andrey2470T What's your status of this? Any plans of continuing it?

(I recently thought about leaving a rough review, or *something*, or ask to help you out. (Because this PR is very important.) But I thought I'd wait first for the more detailed PR description / doc I asked for.)

**Comment by Andrey2470T** [comment]
@Desour I closed it mostly due to the lack of the interest from others since during these recent months there were very little a code review and not enough any feedback generally. And I doubted something would change in this regard. So I'm not sure whether it is really worth to conitnue working on that.

**Comment by sfan5** [comment]
A texture atlas is a big component we will need to improve map rendering performance, we definitely want it.
Like I said I think a good step would be to implement the bare-minimum: an atlas for static textures first and PR that on its own.

**Comment by appgurueu** [comment]
How we ought to reduce drawcalls for different textures is an architectural decision that should be well made. I have not yet arrived at the conclusion that a texture atlas is the way to go (but on the contrary, I'm not experienced enough yet to conclude that it *isn't* the way to go either, so don't consider this a veto or anything; feel free to move forward if you are convinced not only that this solves the right problem, but also that it solves it in a sound way).

I see texture atlases only as one among multiple alternatives, each with its own pros and cons:

* Bindless textures
    * Pros:
      * Aurailus seems to be successfully using a particular bindless texture based implementation in her voxel game engine ([video](https://www.youtube.com/watch?v=YTfdBSjitd8)); she is targeting much more recent hardware than we do though.
      * These are pretty much what we want: The ability to just reference textures from shaders by ID. If integrated well I imagine they should incur minimal maintenance burden.
    * Concerns:
      * It's an extension, not all hardware can profit from this (though I think most moderately recent, and especially gaming hardware, can)
      * devsh: "if you don't have NV_gpu_shader5 then your texture handles need to be ['dynamically uniform'](https://www.khronos.org/opengl/wiki/Core_Language_(GLSL)#Dynamically_uniform_expression)"
* Texture arrays
    * Pros: Seems to have worked well for Build A World (but they could assume textures of the same resolution for all voxels).
    * Cons: As has been pointed out before, they require that textures have the same resolution, but unsurprisingly, many of our textures do (and modders could be encouraged to do this for performance reasons).
* Texture atlases
    * Pros: Most portable, established in many games (especially 2d with handmade atlases, but also 3d); does anyone know a good example of a game engine which autogenerates atlases?
    * Concerns:
       * To what extent do we incur technical debt? Mipmapping *requires* us to tie our meshgen code strongly to this atlas. Textures and meshes can't be as independent anymore.
       * You seem to assume that textures are, in general, just big enough to fit a full atlas containing all textures. I doubt that. You claim a minimum max texture size of 16k squared, but I can't find a source for that. [This source](https://stackoverflow.com/questions/9072764/minimum-required-texture-size-for-compliance-with-opengl-es-2-0-on-android) for example claims a measly min max texture size of 64² on OGLES2, realistically I could imagine something like 1024² or 2048². [This random reddit post](https://www.reddit.com/r/opengl/comments/rqogjw/max_3d_texture_size_value/) also contradicts your assumption. I would also like to add that the assumption that we don't have anything beyond 128² texture packs [doesn't hold either](https://forum.luanti.org/viewtopic.php?t=1583). There also is at least one example of [a game which uses 1024² textures](https://content.luanti.org/packages/Droog71/moontest/). (Which seems excessive to me, but oh well.)
       * According to your testing, there is not yet a profound impact on FPS. How sure are we that future improvements will yield that in conjunction with this? Why can we not do those future improvements first if it is clearer what the right approach is there?
       * To what extent do we use texture wrapping and to what extent would that be broken? How do you avoid seams at tile edges (do you extend texture edges)? Has this been tested against world-aligned textures yet?
       * Mipmapping ultimately *can not* work with a texture atlas. There *will be* a mipmap level at which it will be broken. (And in trying to reduce this level by adding margin, you would have to exponentially blow up your wasted space.) Are we fine with largely throwing mipmapping out the window, and if so what is the cost of this?
       * As has been said already: Mipmapping already failed for this project once. Have we (have you) understood why and how not to repeat the mistakes? (I'm bringing this up again because it seems like at least I am missing a good answer to this so far.)

**TL;DR:** I think further research is needed before committing to this approach, at least for me.

Since it seems likely to me that I might not have time for that soon, I won't hold this PR back based on research I haven't done (and probably won't find time to do), but I encourage whoever wants to commit to it to do that research.

**Comment by Desour** [comment]
> * Texture atlases
>   * Pros:

You can stuff more textures in one atlas than one texture array.
(Which doesn't matter that much if you have only 100 different textures per meshchunk or so.)

>According to your testing, there is not yet a profound impact on FPS.

It does fix the bottleneck, see <https://github.com/minetest/minetest/pull/15061#issuecomment-2360953230>.

>Mipmapping ultimately can not work with a texture atlas. There will be a mipmap level at which it will be broken. (And in trying to reduce this level by adding margin, you would have to exponentially blow up your wasted space.) Are we fine with largely throwing mipmapping out the window, and if so what is the cost of this?

Idk if you're aware of this, but one can set a maximum mipmap level. (And IIRC, this PR does this.)

If all textures in the atlas have the same size, you get a 9x memory overhead (note: this is not an issue if mem usage is dominated by other things, like vertex data). And if they are not all the same size, idk how you'd do it without scaling up. Why exponentially, and exponentially in what???

And no, we're not fine with throwing mipmapping out of the window. I don't want that aliasing.

**Comment by sfan5** [comment]
Mipmapping is a good point. Have you tested it and what's your solution? @Andrey2470T 
I think you could avoid problems by padding textures to powers-of-two and only sharing an atlas between textures of the same size *and* setting the max mipmap level appropriately.

> If all textures in the atlas have the same size, you get a 9x memory overhead 

can you elaborate?

> You can stuff more textures in one atlas than one texture array.

sort of but I don't think this will be a problem for us:
> However, the dimension that represents the number of array levels is limited by GL_MAX_ARRAY_TEXTURE_LAYERS, which is generally much smaller. OpenGL 4.5 requires that the limit be at least 2048, while OpenGL 3.0 requires it be at least 256. 

(https://www.khronos.org/opengl/wiki/Array_Texture)

**Comment by rubenwardy** [comment]
I'm surprised this doesn't have a big impact on FPS, has someone run it though a profiler like renderdoc?

(Sorry, I don't know render stuff that well)

**Comment by AbduSharif** [comment]
> I'm surprised this doesn't have a big impact on FPS
A bit too early to tell tbh, not much testinf has been done in general and it doesn't yet work on Android (which is one of the most likely hardware that would benefit from this), it at least improved a performance bottleneck as Desor mentioned:

> It does fix the bottleneck, see https://github.com/minetest/minetest/pull/15061#issuecomment-2360953230.

**Comment by AbduSharif** [comment]
I have an old PC I'm particularly interested in testing this with,  especially with texturepacks.

**Comment by AbduSharif** [comment]
On that old PC, I have Intel GMA X4500 and the game crashes immediatly with the PR, The crash dump is 0 bytes, There's a log message though:
>  WARNING[Main]: Irrlicht: Could not create ITexture, texture needs to have a non-empty name.

That IGPU's max GL version is 2.1, I assume it doesn't support this feature? What were the requirements?

**Comment by AbduSharif** [comment]
For research purposes :) , on fairly recent llvmpipe for Windows, the textures are black and white:

minetest__atlas_llvmpipe

Something to note though, since what on this PC is a fairly old CPU being used, not all the CPU extensions llvmpipe could make use of are supported.

Going to try on linux next.

**Comment by Andrey2470T** [comment]
I've done a little performance testing with this PR (actually I did that already at the same branch tip a few months ago, but just didn't publish them yet). My videocard is GeForce GTX 1050. I've compared the performance on the current master branch and on my atlas PR's branch tip.

### viewing_range = 190, max_fps=500, driver=opengl3, all shader effects=off, no water around, only solid mapblocks

Vanilla Luanti-5.11.0-dev
Снимок экрана от 2024-11-21 00-02-26

Minetest-5.10.0-dev with atlas
Снимок экрана от 2024-11-21 00-03-03

### viewing_range = 190, max_fps=500, driver=opengl3, all shader effects=off, already with transparent mapblocks (water)

Vanilla Luanti-5.11.0-dev
Снимок экрана от 2024-11-21 00-05-09

Minetest-5.10.0-dev with atlas
Снимок экрана от 2024-11-21 00-05-34

### viewing_range = 450, max_fps=500, driver=opengl3, all shader effects=off, from the same place as at the previous one

Vanilla Luanti-5.11.0-dev
Снимок экрана от 2024-11-21 00-08-54

Minetest-5.10.0-dev with atlas
Снимок экрана от 2024-11-21 00-08-14

**Comment by Desour** [comment]
>> If all textures in the atlas have the same size, you get a 9x memory overhead
>
> can you elaborate?

On the smallest level, each texture (assumed to be square and power of 2) is one pixel big. You need a pixel with the same color on each side, to avoid bleeding issues. It's 9x as big on the lowest level, so it's 9x as big on each level, and hence 9x as big overall. (Well, it's + 8x overhead, to be exact.)
(Because of different wrapping modes, one might have to have duplicates. I've ignored this here.)
(This overhead could be made smaller (by a factor of 4) if one chooses the smallest level to be 2x2 pixels, in exchange for aliasing.)

**Comment by sfan5** [comment]
> You need a pixel with the same color on each side, to avoid bleeding issues

I see, forgot about that.

> Testing this in almost empty scenes, where you get >60 fps anyways, make this look less important than it is, btw..

Indeed. It helps most on scenes with lots of different nodes (textures).

**Comment by Zughy** [comment]
@Andrey2470T any updates? Also rebase

**Comment by sfan5** [comment]
wanted to note that I think array textures are the way to go:
https://irc.luanti.org/luanti-dev/2025-03-30#i_6252312
